enhancements/20240612-node-disk-manager-refactor.md

```diff
@@ -0,0 +1,67 @@
+# Node Disk Manager Refactor
+
+## Summary
+
+The Node Disk Manager is responsible for managing the raw device on the node. The current implementation is not flexible enough to support various storage backends. This enhancement is to refactor the Node Disk Manager to make it more flexible and extensible.
+
+### Related Issues
+
+https://github.com/harvester/harvester/issues/6014
+
+## Motivation
+
+We need to refactor the current implementation to make the Node Disk Manager support various storage backends. This could help define the storage provisioner easily.
+
+### Goals
+
+Refactor the Node Disk Manager to make it more flexible and extensible to support various storage backends. The refactor will decompose the individual storage provisioner.
+
+### Non-goals [optional]
+
+None
+
+## Proposal
+
+The Node Disk Manager refactor will mainly focus on the provision. It will introduce the provisioner concept to define the specific storage provisioner. The provisioner will format the raw device and then provision it to the corresponding storage backends.
+
+### User Stories
+
+#### The user wants to use the longhornV2 engine
+The user wants to try the LonghornV2 engine on the Harvester. The user can use the longhornv2 provisioner to provision the raw device to Longhorn for the v2 engine.
+
+#### The user wants to use LVM as the storage backend
+The user wants to use LVM as the storage backend for better performance. The node-disk-manager would help create the volume group, and the LVM CSI driver could depend on it to create the logical volume.
+
+### User Experience In Detail
+
+The user could use a specific provisioner like `longhornv1`, `longhornv2`, or `LVM` to provision the raw device to the corresponding storage backend. After that, we could easily create PVC from the specific storage backend.
+
+### API changes
+
+TBD
+
+## Design
+
+### Implementation Overview
+
+A overview of the design could be checked below:
+![ndm-refactor](20240612-node-disk-manager-refactor/ndm-refactor.png)
+
+### Provisioner
+
+The refactor will try to decompose the provisioner from the controller. Each provisioner will be responsible for the specific storage backend. For example, the longhornv1 provisioner will handle the filesystem formatting on the Format stage and provision this raw device to `nodes.longhorn.io`. That longhorn could create volume on the provisioned raw device. The LVM provisioner only handles the provision stage to create the volume group.
+
+### Test plan
+
+1. Create Harvester cluster
+2. Create supported storage provisioner (e.g. longhornv1, longhornv2, and LVM)
```

Comment:
I'd suggest make a fake provisioner to do unit test. It shows that you create a well interface that can be replaced by any new provisioners.

Reply:
Thanks for the suggestion.
I added your suggestion on the unit test part of the design.

I thought the `Test plan` was used for integration tests or QA validation, so I prefer to keep that.
WDYT?

Reply:
Oh, you're right. I misunderstood it. Sorry for that. I don't have problem now. Thanks!